src/Juvix/Backends/ArithmeticCircuit/Compilation/Types.hs

```diff
@@ -1,19 +1,183 @@
 module Juvix.Backends.ArithmeticCircuit.Compilation.Types where
 
-import qualified Data.Graph.Inductive as G
-import Juvix.Library
-
-type Circuit = G.Gr NodeType WireType
-
-data NodeType
-  = -- Addition gate.
-    Add
-  | -- Multiplication gate.
-    Mul
-  | -- Input value (by vector index).
-    InputValue Natural
-  | -- Output value (by vector index).
-    OutputValue Natural
-  deriving (Show, Eq, Generic)
-
-type WireType = ()
+import qualified Juvix.Core.ErasedAnn as CoreErased
+import qualified Juvix.Core.Usage as Usage
+import qualified Juvix.Library as J
+import qualified Juvix.Backends.ArithmeticCircuit.Parameterisation as Par
+import qualified Circuit.Expr as Expr
+import qualified Circuit.Lang as Lang
+import qualified Circuit
+import qualified Data.Map as Map
+import Numeric.Natural()
+
+import Protolude
+
+data PrimVal
+  = Element Par.F
+  | Boolean Bool
+  | FEInteger Int
+  | BinOp BinOp Term Term
+  | Op Op Term
+  | If Term Term Term
+  deriving (Show, Eq)
+
+data BinOp
+  = Add
+  | Mul
+  | Sub
+  | Exp
+  | Eq
+  | And
+  | Or
+  deriving (Show, Eq)
+
+data Op = Neg
+  deriving (Show, Eq)
+
+type Term = CoreErased.AnnTerm () PrimVal
+type Type = CoreErased.Type () PrimVal
+
+data CompilationError
+  = NotYetImplemented
+  | SomethingWentWrongSorry
+  | VariableOutOfScope
+  deriving (Eq, Show)
+
+data ArithExpression f
+  = BoolExp (Expr.Expr Circuit.Wire f Bool)
+  | FExp (Expr.Expr Circuit.Wire f f)
+  | NoExp
+
+data Memory = Mem (Map.Map J.Symbol (Int, ArithExpression Par.F)) Int
+
+instance Semigroup Memory where
+  (Mem map_ n) <> (Mem map__ m) = Mem (map_ <> Map.map (shift n) map__) (n + m)
+    where shift n (m, exp) = (n + m, exp)
+
+instance Monoid Memory where
+  mempty = Mem Map.empty 0
+  mappend = (<>)
+
+insert :: J.Symbol -> ArithExpression Par.F -> Memory -> Memory
+insert sy x (Mem map_ n) = Mem (Map.insert sy (n, x) map_) (n + 1)
+
+lookup :: J.Symbol -> Memory -> Maybe (Int, ArithExpression Par.F)
+lookup sy (Mem map_ _) =
+  case Map.lookup sy map_ of
+    Just res -> Just res
+    Nothing -> Nothing
+
+transTerm :: Memory
```

Comment:
IMO this would be cleaner in `MonadState Memory`

Reply:
Ditto, it'd make the trans functions lcleaer cleaner as well